### PR TITLE
Added sync2.Set

### PR DESCRIPTION
**Pull request by Alexamakans**
Thread-safe set.

### CODE REVIEW

**Review by jilleJr — COMMENTED** [review]
While I love the Initiative, I have some problems with this implementation:

- Mix of responsibility for the sync2 package, especially considering the copied Product type and CartesianProduct function. The code better belongs inside the sets package, and given the added set type then maybe sets.Set should be an interface? Just a thought.

- There's no good interop with sets.Set. There's a wrap method, but that's pretty clunky.

- Would rather see the implementation make use of sync2.Map instead, as it already contains all the complex logic for having a lock-free map.

- I've strictly been staying away from dependencies for this repo, even in testing, because I want this repo to be a dependency-free implementation where it's completely standalone. It's a silly constraint, but I want to stick with it.

To add this, maybe a major bump is needed given my idea of how this could fit in.

I'll revisit this, but I'm just leaving these comments here for now. I'm sadly not willing to merging this as-is

**Comment by Alexamakans** [comment]
> Would rather see the implementation make use of sync2.Map instead, as it already contains all the complex logic for having a lock-free map.

Do those complexities apply when it's used as a set?
Is it mainly complex because of the dirty/readonly portion of it?

Haven't looked at it a whole lot nor thought about the issue, will check it out.

> I've strictly been staying away from dependencies for this repo, even in testing, because I want this repo to be a dependency-free implementation where it's completely standalone. It's a silly constraint, but I want to stick with it.

I imagined this was the case when I saw there were none, I see now that it's written in the `README.md` as well.
I like this constraint 👍🏻

Will look at doing it without dependencies.

> - Mix of responsibility for the sync2 package, especially considering the copied `Product` type and `CartesianProduct` function. The code better belongs inside the sets package, and given the added set type then maybe `sets.Set` should be an interface? Just a thought.
> - There's no good interop with `sets.Set`. There's a wrap method, but that's pretty clunky.

Both good points.
In addition to maybe using `sync2.Map` instead, I will also likely look at alternative ways to make it fit better in `pkg/sets`.

**Comment by jilleJr** [comment]
Ok I got a plan:

- move sets.Set to maps.Set, as it's a set implementation based on the map

- keep this as sync2.Set, as it does live nicely next to sync2.Map

- change it to use sync2.Map for the implementation.

- do a major version bump.

- make CartesianProduct function work on interfaces instead so it can be used on both sync2.Set and maps.Set.

- probably rename some methods along the way, as I at least still always confuse SetDiff with SymDiff.

Do you want to give this a shot, or shall I apply my proposed changes?

**Comment by Alexamakans** [comment]
Sounds good.
I'll give it a shot 👍🏻

File: pkg/maps/set_example_test.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	fmt.Println("intersect:", set1.Intersect(set2)) // {B C}
+	fmt.Println("set diff:", set1.SetDiff(set2))    // {A}
+	fmt.Println("sym diff:", set1.SymDiff(set2))    // {A D}
+
+	// Please note: the Set.String() output is not sorted!
+}
+
+func TestSet_SetOperations(t *testing.T) {
+	setA := newSetA()
+	assert.ElementsMatch(t, setA.Slice(), []string{"A", "B", "C"}, "setA values")
```
Comment (Alexamakans) [inline pkg/maps/set_example_test.go:52]: Any tip for built in version of elements match or should I just write my own?

Comment (jilleJr) [reply pkg/maps/set_example_test.go:52]:
```go
want := "A, B, C"
got := strings.Join(setA.Slice(), ", ")
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,14 @@ SPDX-License-Identifier: CC-BY-4.0
 
 This project tries to follow [SemVer 2.0.0](https://semver.org/).
 
+## v4.0.0 (WIP)
+
+- Added interface `sets.Set`. (#?)
```
Comment (jilleJr) [inline CHANGELOG.md:15]:
```suggestion
- Added interface `sets.Set`. (#32)
```

File: pkg/maps/set_example_test.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	fmt.Println("union:", set1.Union(set2))         // {A B C D}
+	fmt.Println("intersect:", set1.Intersect(set2)) // {B C}
+	fmt.Println("set diff:", set1.SetDiff(set2))    // {A}
+	fmt.Println("sym diff:", set1.SymDiff(set2))    // {A D}
+
+	// Please note: the Set.String() output is not sorted!
+}
+
+func TestSet_SetOperations(t *testing.T) {
+	setA := newSetA()
```
Comment (jilleJr) [inline pkg/maps/set_example_test.go:51]: This is a file with only example tests. Suggest moving these test functions to a new file `set_test.go`

File: pkg/maps/set_example_test.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+func TestSet_String(t *testing.T) {
+	set1 := make(maps.Set[string], 0)
+	set1.Add("A")
+	assert.Equal(t, "{A}", set1.String(), "one value")
+
+	set1.Add("B")
+	str := set1.String()
+	assert.True(t, str == "{A B}" || str == "{B A}", "two values")
```
Comment (jilleJr) [inline pkg/maps/set_example_test.go:101]: Please use `assert.Comparable` from `internal/assert` instead

File: pkg/sets/sets.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+	result := make(map[Product[TA, TB]]struct{}, 0)
+	a.Range(func(valueA TA) bool {
+		b.Range(func(valueB TB) bool {
+			result[Product[TA, TB]{valueA, valueB}] = struct{}{}
+			return true
+		})
+		return true
+	})
+	return result
+}
```
Comment (jilleJr) [inline pkg/sets/sets.go:107]: Suggest returning a slice of products instead.

File: pkg/slices/slices.go [hunk 6]
```diff
@@ -444,13 +444,13 @@ func ChunkFunc[S ~[]E, E any](slice S, size int, callback func(chunk S)) {
...
 func Except[S ~[]E, E comparable](slice S, exclude S) S {
-	set := sets.NewFromSlice(exclude)
+	set := maps.NewSetFromSlice(exclude)
 	return ExceptSet(slice, set)
 }
 
 // ExceptSet returns a new slice for all items that are not found in the set of
 // items to exclude.
-func ExceptSet[S ~[]E, E comparable](slice S, exclude sets.Set[E]) S {
+func ExceptSet[S ~[]E, E comparable](slice S, exclude maps.Set[E]) S {
```
Comment (jilleJr) [inline pkg/slices/slices.go:453]: Should be able to use `sets.Set` here right?

File: pkg/sync2/set.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+// Add will add an element to the set, and return true if it was added
+// or false if the value already existed in the set.
+func (s *Set[T]) Add(value T) bool {
+	if s.Has(value) {
+		return false
+	}
+	s.m.Store(value, struct{}{})
+	return true
```
Comment (jilleJr) [inline pkg/sync2/set.go:78]: Possible race condition where the value was added elsewhere between the Has and Store

Suggest using [LoadOrStore](https://pkg.go.dev/sync#Map.LoadOrStore) instead

```suggestion
	_, loaded := s.m.LoadOrStore(value, struct{}{})
	return !loaded
```

File: pkg/sync2/set.go [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+}
+
+// Remove will remove an element from the set, and return true if it was removed
+// or false if no such value existed in the set.
+func (s *Set[T]) Remove(value T) bool {
+	if !s.Has(value) {
+		return false
+	}
+	s.m.Delete(value)
+	return true
```
Comment (jilleJr) [inline pkg/sync2/set.go:101]: Possible race condition, should use [LoadAndDelete](https://pkg.go.dev/sync#Map.LoadAndDelete) instead

```suggestion
	_, loaded := s.m.LoadAndDelete(value)
	return loaded
```

File: pkg/sync2/set.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+// Clone returns a copy of the set.
+func (s *Set[T]) Clone() sets.Set[T] {
+	var clone Set[T]
+	clone.AddSet(s)
+	return &clone
+}
+
+// Slice returns a new slice of all values in the set.
+func (s *Set[T]) Slice() []T {
+	result := make([]T, 0)
```
Comment (jilleJr) [inline pkg/sync2/set.go:126]: No need to initialize the slice if you're setting it to zero anyway

```suggestion
	var result []T
```

File: pkg/sync2/set_example_test.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+	fmt.Println("set diff:", set1.SetDiff(set2))    // {A}
+	fmt.Println("sym diff:", set1.SymDiff(set2))    // {A D}
+
+	// Please note: the Set.String() output is not sorted!
+}
+
+func TestSet_SetOperations(t *testing.T) {
+	setA := newSetA()
+	assert.ElementsMatch(t, setA.Slice(), []string{"A", "B", "C"}, "setA values")
+	setB := newSetB()
```
Comment (jilleJr) [inline pkg/sync2/set_example_test.go:53]: Please move unit tests out from the example tests file

File: pkg/sync2/set_example_test.go [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+	}
+}
+
+func newSetA() sets.Set[string] {
+	setA := &sync2.Set[string]{}
+	setA.Add("A")
+	setA.Add("B")
+	setA.Add("C")
+	return setA
+}
```
Comment (jilleJr) [inline pkg/sync2/set_example_test.go:134]: Suggest renaming these functions to

- newSetABC
- newSetBCD

for the extra clarity

File: pkg/sync2/set_example_test.go [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+func TestSet_SetOperations(t *testing.T) {
+	setA := newSetA()
+	assert.ElementsMatch(t, setA.Slice(), []string{"A", "B", "C"}, "setA values")
+	setB := newSetB()
+	assert.ElementsMatch(t, setB.Slice(), []string{"B", "C", "D"}, "setB values")
+
+	assert.ElementsMatch(t, setA.Union(setB).Slice(), []string{"A", "B", "C", "D"}, "union")
+	assert.ElementsMatch(t, setA.Intersect(setB).Slice(), []string{"B", "C"}, "intersect")
+	assert.ElementsMatch(t, setA.SetDiff(setB).Slice(), []string{"A"}, "set diff")
+	assert.ElementsMatch(t, setA.SymDiff(setB).Slice(), []string{"A", "D"}, "sym diff")
```
Comment (jilleJr) [inline pkg/sync2/set_example_test.go:59]: Please use one unit test function per method tested

**Review by jilleJr — CHANGES_REQUESTED** [review]

**Comment by jilleJr** [comment]
Regarding the REUSE compliance, I suggest:

- Files you've created completed from scratch, add your own copyright, e.g:

  ```go
  // SPDX-FileCopyrightText: 2022 <your name>
  //
  // SPDX-License-Identifier: MIT
  ```

- Files you've taken based on my files, add your copyright to the same that was in the forked file, e.g:

  ```go
  // SPDX-FileCopyrightText: 2022 <your name>
  // SPDX-FileCopyrightText: 2022 Kalle Fagerberg
  //
  // SPDX-License-Identifier: MIT
  ```

As I license this under MIT, I require that you're OK with licensing your code under MIT as well

File: pkg/maps/set_example_test.go (hunk 1 above)
Comment (Alexamakans) [reply pkg/maps/set_example_test.go:52]: Considering it's based on maps and is unordered, anything else? Not that it's a complicated thing to write for this case, but if there's something built in that'd be nice.

Comment (jilleJr) [reply pkg/maps/set_example_test.go:52]: Nothing built in, but you could write a function for it in `internal/assert`

Comment (jilleJr) [reply pkg/maps/set_example_test.go:52]: Such as:

```go
func ElementsMatch[S ~[]string](t *testing.T, want, got S) {
    want := slices.Clone(want)
    got := slices.Clone(got)
    slices.Sort(want)
    slices.Sort(got)
    wantStr := strings.Join(want, ", ")
    gotStr := strings.Join(got, ", ")
    if wantStr != gotStr {
        t.Errorf("want %q, got %q", wantStr, gotStr)
    }
}
```

File: pkg/slices/slices.go (hunk 6 above)
Comment (Alexamakans) [reply pkg/slices/slices.go]: Yup 👍🏻

**Comment by Alexamakans** [comment]
> Regarding the REUSE compliance, I suggest:
> 
> * Files you've created completed from scratch, add your own copyright, e.g:
>   ```go
>   // SPDX-FileCopyrightText: 2022 <your name>
>   //
>   // SPDX-License-Identifier: MIT
>   ```
> * Files you've taken based on my files, add your copyright to the same that was in the forked file, e.g:
>   ```go
>   // SPDX-FileCopyrightText: 2022 <your name>
>   // SPDX-FileCopyrightText: 2022 Kalle Fagerberg
>   //
>   // SPDX-License-Identifier: MIT
>   ```
> 
> As I license this under MIT, I require that you're OK with licensing your code under MIT as well

Not sure I have done this for every applicable file, but I think I got them all.

File: internal/assert/assert.go [hunk 13]
```diff
@@ -4,11 +4,32 @@
...
 }
+
+// ElementsMatch asserts that two slices of strings contain the same values,
+// with no regard to the ordering.
+func ElementsMatch[S ~[]string](t *testing.T, want, got S) {
+	var wantClone S
+	copy(wantClone, want)
+	sort.Strings(wantClone)
+	var gotClone S
+	copy(gotClone, got)
```
Comment (jilleJr) [inline internal/assert/assert.go:27]: This doesn't work. The `copy` function only copies as many elements as fits, and a nil slice has a length of 0, meaning both of these `copy` operations result in two empty slices. In effect, this function doesn't assert anything :)

You need to initialize their sizes, such as like so:

```suggestion
	wantClone := make(S, len(want))
	copy(wantClone, want)
	sort.Strings(wantClone)
	gotClone := make(S, len(got))
	copy(gotClone, got)
```

**Comment by Alexamakans** [comment]
The duplication it's complaining about is solely the test code for `sync2.Set` and `maps.Set`.
Make it nicer, or is this acceptable?

**Comment by jilleJr** [comment]
> The duplication it's complaining about is solely the test code for `sync2.Set` and `maps.Set`. Make it nicer, or is this acceptable?

Acceptable.

File: pkg/maps/set.go [hunk 14]
```diff
@@ -1,44 +1,46 @@
+// SPDX-FileCopyrightText: 2022 Per Alexander Fougner
```
Comment (jilleJr) [inline pkg/maps/set.go:1]: Per? 🤭 How official of you :)

File: internal/assert/assert.go (hunk 13 above)
Comment (Alexamakans) [reply internal/assert/assert.go:27]: Aah, gotcha 👍🏻

File: pkg/maps/set.go (hunk 14 above)
Comment (Alexamakans) [reply pkg/maps/set.go:1]: There's another Swedish programmer by the name of Alexander Fougner hehe

**Review by jilleJr — APPROVED** [review]
Good job! Thank you for the extensive follow-up!

**Pull request merged**